### DISCUSSION TITLE
Boolean indexing alternative

### DISCUSSION THREAD

**Discussion by Kheims** [ml-explore/mlx#408] [Q&A] [answered]
Hi, I have seen that boolean indexing isn't yet implemented.

Particularly, I have seen @awni's workaround recommandation ( [#246](https://github.com/ml-explore/mlx/issues/246)) on using `mx.where`, but apparently it has issues handling inf values for x (it's a TODO in the cpp operation), which is the case for us.

Here is our usecase : 

```
        if normalize:
            deg = self._degree(col, x.shape[0])
            deg_inv_sqrt = deg ** (-0.5)
            # NOTE : need boolean indexing in order to zero out inf values 
            deg_inv_sqrt[deg_inv_sqrt == float('inf')] = 0 # doesn't work yet
            deg_inv_sqrt = mx.where(deg_inv_sqrt == float('inf'), mx.zeros_like(deg_inv_sqrt), deg_inv_sqrt) # Nan in place of zeros for inf values
            norm = deg_inv_sqrt[row] * deg_inv_sqrt[col]
```

Is there an alternative to using numpy or worse a listcomp ?

**Comment by awni** [comment]
That's a tricky one 🤔 . 

Of course we need to fix the where to handle the `inf` case. That is not conceptually difficult, but might require a primitive so it's a bit of work.

Just curious why does `deg` have `inf` in it? Is it a sentinal for something? If so could you use a different value? Like maybe just something very large?

**Reply by Kheims** [reply]
deg has inf because it contains the degree (ie `#of neighbouring nodes`) of each node in a graph data struct, so some of them may be isolated.
It might be the case especially when batching large graphs. or when dealing with directed graphs

**Reply by awni** [reply]
I see so `deg=0` in some cases, makes sense.

**Reply by awni** [reply]
How about you flip your computation in that case:

```python
deg = self._degree(col, x.shape[0])
deg_inv_sq = mx.where(deg == 0, float("inf"), deg) ** (-0.5)
```

**Reply by awni** [reply]
PS the reason the `where` has a NaN is because `inf * 0 = NaN`. So another simple alternative is to switch the cases the way you have it. Something like this:

```python
deg_inv_sqrt = mx.where(deg_inv_sqrt < float('inf'), deg_inv_sqrt, 0) 
```

**Reply by Kheims** [reply]
Both solutions didn't solve the issue unfortunately :(.
```
a = mx.array([1,2,3, float('inf')])
mx.where(a < float('inf'), a, 0) # output : [1,2,3, nan]
```

**Reply by awni** [reply]
You're right disregard those, sorry they are not correct, the NaNs just get moved..

**Reply by Kheims** [reply]
I guess the viable solution for now is to create the indices with numpy ? But we for sure will need a consistent solution as this will be an important step in GCNConv's forward.
If you can point me to the right direction i'd gladly do my best to contribute to the where operation.

**Reply by awni** [reply, marked as answer]
@Kheims It shouldn't be too bad to precompute the non-zero degree indices from a perf standpoint. I know it is not that elegant.

A slightly hacky solution could be to do something like:

```shell
>>> import mlx.core as mx
>>> deg = mx.array([1,2,3,0])
>>> deg += 1e-10
>>> deg_inv_sq = deg**(-0.5)
>>> deg_inv_sq = mx.where(deg_inv_sq <= 1, deg_inv_sq, 0)
>>> deg_inv_sq
array([1, 0.707107, 0.57735, 0], dtype=float32)
```

**Reply by Kheims** [reply]
Thank you for the workaround :)
